src/prototype_rust_sdk/workflow_future.rs

```diff
@@ -266,8 +267,28 @@ impl Future for WorkflowFuture {
                 return Ok(WfExitValue::Evicted).into();
             }
 
-            // TODO: Trap panics in wf code here?
-            let mut res = self.inner.poll_unpin(cx);
+            // TODO: Make sure this is *actually* safe before un-prototyping rust sdk
+            let mut res = match AssertUnwindSafe(&mut self.inner)
+                .catch_unwind()
+                .poll_unpin(cx)
+            {
+                Poll::Ready(Err(e)) => {
+                    self.outgoing_completions
+                        .send(WfActivationCompletion::fail(
+                            &self.task_queue,
+                            run_id,
+                            Failure {
+                                message: format!("Workflow function panicked: {:?}", e),
+                                ..Default::default()
+                            },
+                        ))
+                        .expect("Completion channel intact");
+                    // Loop back up because we're about to get evicted
```

Comment:
Pardon my ignorance to the algorithm, can you help me understand the flow and this comment? Who's listening to `WfActivationCompletion`? Are they then expected to send back a `WfActivation` with an eviction job upon receipt? Is it expected that there is only one `WfActivationCompletion` per workflow and if so, what if this completion is sent _and_ when the incoming eviction is processed another completion is sent?

Reply:
This would be easier to explain on a quick call. The actual core library eventually gets the completion. I know it will send an eviction activation for this workflow because the lang SDK (which just happens to be Rust b/c test) just sent a failure.

There can only ever be one outstanding activation for a given workflow run at a time, so the race you describe is not possible.

Feel free to put something on my calendar and I can go over this in more detail